src/helpers.ts

```diff
@@ -136,3 +156,145 @@ export async function addTeamLabel(context: Context, octokit: OctokitType, pullR
         labels: [teamLabelName],
     });
 }
+
+/**
+ * Sends a query to ZenHub GraphQL API server using Axios client.
+ */
+async function queryZenhubGraphql(operationName: string, query: string, variables: object) {
+    const zenhubToken = core.getInput('zenhub-token');
+
+    return axios({
+        method: 'post',
+        url: 'https://api.zenhub.com/public/graphql',
+        headers: {
+            Authorization: `Bearer ${zenhubToken}`,
+        },
+        data: { query, variables, operationName },
+    });
+}
+
+const ZENHUB_PR_DETAILS_QUERY = `
+query getIssueInfo($repositoryGhId: Int!, $issueNumber: Int!) {
+    issueByInfo(repositoryGhId: $repositoryGhId, issueNumber: $issueNumber) {
+        id
+        repository {
+            id
+            ghId
+        }
+        number
+        title
+        body
+        state
+        timelineItems(first: 100) {
+            nodes {
+                type: key
+                id
+                data
+                createdAt
+            }
+        }
+        estimate {
+            value
+        }
+    }
+}
+`;
+
+const ZENHUB_ISSUE_ESTIMATE_QUERY = `
+query getIssueInfo($repositoryGhId: Int!, $issueNumber: Int!) {
+    issueByInfo(repositoryGhId: $repositoryGhId, issueNumber: $issueNumber) {
+        estimate {
+            value
+        }
+    }
+}
+`;
+
+/**
+ * Makes sure that:
+ * - PR either has issue or epic linked or has `adhoc` label
+ * - either PR or linked issue has estimate
+ */
+export async function ensureCorrectLinkingAndEstimates(pullRequest: PullRequest, octokit: OctokitType): Promise<void> {
+    const pullRequestGraphqlResponse = await queryZenhubGraphql('getIssueInfo', ZENHUB_PR_DETAILS_QUERY, {
+        repositoryGhId: pullRequest.head.repo?.id,
+        issueNumber: pullRequest.number,
+        workspaceId: ZENHUB_WORKSPACE_ID,
+    });
+
+    const pullRequestEstimate = pullRequestGraphqlResponse.data.data.issueByInfo.estimate?.value;
+    const linkedIssue = getLinkedIssue(pullRequestGraphqlResponse.data.data.issueByInfo.timelineItems.nodes);
+    const linkedEpics = getLinkedEpics(pullRequestGraphqlResponse.data.data.issueByInfo.timelineItems.nodes);
+
+    if (
+        !linkedIssue
+        && linkedEpics.length === 0
+        && !pullRequest.labels.some(({ name }) => name === 'adhoc')
+    ) await fail(pullRequest, `⚠️ [Pull Request Tookit](https://github.com/apify/pull-request-toolkit-action) has failed!\n\n> Pull request is neither linked to an issue or epic nor labeled as adhoc!`, octokit);
```

Comment:
I would move this `"⚠️ [Pull Request Tookit](https://github.com/apify/pull-request-toolkit-action) has failed!\n\n>" ` part to the `fail` function, so that it's in all the `fail()` messages.